Fix Deep Scanline Input crash when using a framebuffer as parameter

nikos-foundry opened the pull request:
The Deep Scanline Input code has recently been rewritten to make use of the EXR core library. As a result a bug was introduced when using the overloads of readPixels() and readPixelSampleCounts() that use externally provided framebuffers as parameters, a hard crash was occuring. According to the comment on the header of these functions, when these overloads are used the InputFile's frameBuffer is not used.
However both these functions call into readMemData(), which does use the InputFile's member framebuffer, which has not been allocated, resulting into a segmentation fault.
This commit fixes the above issue, by using the parameter framebuffer that is passed in, which should be allocated by the caller application.

Closes https://github.com/AcademySoftwareFoundation/openexr/issues/2020

nikos-foundry committed:
Fix Deep Scanline Input crash when using an externally allocated framebuffer

The Deep Scanline Input code has recently been rewritten to make use of the
EXR core library. As a result a bug was introduced when using the overloads of
readPixels() and readPixelSampleCounts() that use externally provided
framebuffers as parameters, a hard crash was occuring. According to the comment
on the header of these functions, when these overloads are used the InputFile's
frameBuffer is not used.
However both these functions call into readMemData(), which does use the
InputFile's member framebuffer, which has not been allocated, resulting into a
segmentation fault.
This commit fixes the above issue, by using the parameter framebuffer that is
passed in, which should be allocated by the caller application.

Signed-off-by: Nikolaos Koutsikos <nikolaos.koutsikos@foundry.com>

lgritz commented:
Does this reveal a hole in our testing? Is there a test that can be added to validate the fix? Does the same bug exist for deep tiled input?

nikos-foundry commented:
> Does this reveal a hole in our testing? Is there a test that can be added to validate the fix? Does the same bug exist for deep tiled input?

Yes, there seems to be a hole in testing, since these overloaded functions are not being used in any of the existing tests. I will give it a go to add a case to one of the existing tests that uses these functions to extend our coverage.

I think that DeepTiledInput is not affected by this issue, since it seems to be using a different code path, and does not seem to use the problematic function (readMemData).

nikos-foundry committed:
Add a Deep Scanline Input test that uses a framebuffer parameter

After having found a bug with the readPixels() and readPixelSampleCounts()
overloads that receive a framebuffer parameter, it was made apparent that there
was a hole in our tests to cover these functions.
This commits slightly refactors the DeepScanLineBasic tests and adds a third
read path to add coverage for this case as well.

Signed-off-by: Nikolaos Koutsikos <nikolaos.koutsikos@foundry.com>

peterhillman reviewed (approved):
Looks good! Thanks for the fix to the library and to the test.
